File: src/hyperhandler/models/order.py
"""Order-related models."""

from dataclasses import dataclass, field
from datetime import datetime
from decimal import Decimal
from enum import Enum


class OrderStatus(str, Enum):
    """Order status enum."""

    PENDING = "pending"
    OPEN = "open"
    FILLED = "filled"
    PARTIALLY_FILLED = "partially_filled"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


@dataclass
class OrderResult:
    """Result of an order placement."""

    success: bool
    order_id: int | None = None
    filled_size: Decimal = field(default_factory=lambda: Decimal("0"))
    avg_price: Decimal | None = None
    error: str | None = None
    status: OrderStatus = OrderStatus.PENDING

    @property
    def is_filled(self) -> bool:
        """Check if order is fully filled."""
        return self.status == OrderStatus.FILLED

    @property
    def is_partial(self) -> bool:
        """Check if order is partially filled."""
        return self.status == OrderStatus.PARTIALLY_FILLED


@dataclass
class OpenOrder:
    """Representation of an open order."""

    coin: str
    order_id: int
    side: str  # "B" for buy, "S" for sell
    price: Decimal
    size: Decimal
    timestamp: int
    order_type: str = "limit"
    reduce_only: bool = False

    @property
    def is_buy(self) -> bool:
        """Check if this is a buy order."""
        return self.side == "B"


@dataclass
class Position:
    """Representation of an open position."""

    coin: str
    size: Decimal  # Positive for long, negative for short
    entry_price: Decimal
    position_value: Decimal
    unrealized_pnl: Decimal
    leverage: int
    leverage_type: str  # "cross" or "isolated"
    liquidation_price: Decimal | None = None

    # Risk tracking fields
    mark_price: Decimal | None = None
    funding_accrued: Decimal = field(default_factory=lambda: Decimal("0"))
    stop_loss_price: Decimal | None = None
    opened_at: datetime | None = None

    # Calculated risk fields (populated by RiskManager)
    risk_amount: Decimal | None = None  # $ at risk = size * |entry - stop|
    risk_pct: Decimal | None = None  # % of account value
    correlation_group: str | None = None  # "btc-major", "l1-alt", etc.

    @property
    def is_long(self) -> bool:
        """Check if this is a long position."""
        return self.size > 0

    @property
    def is_short(self) -> bool:
        """Check if this is a short position."""
        return self.size < 0

    @property
    def abs_size(self) -> Decimal:
        """Get absolute position size."""
        return abs(self.size)

File: internal/models/order.go
package models

import (
	"time"

	"github.com/shopspring/decimal"
)

// OrderStatus is the lifecycle state of an order.
type OrderStatus string

const (
	StatusPending         OrderStatus = "pending"
	StatusOpen            OrderStatus = "open"
	StatusFilled          OrderStatus = "filled"
	StatusPartiallyFilled OrderStatus = "partially_filled"
	StatusCancelled       OrderStatus = "cancelled"
	StatusRejected        OrderStatus = "rejected"
)

// OrderResult is the outcome of an order placement. Mirrors order.py:OrderResult.
type OrderResult struct {
	Success    bool
	OrderID    *int64
	FilledSize decimal.Decimal
	AvgPrice   *decimal.Decimal
	Error      string
	Status     OrderStatus
}

// IsFilled reports whether the order is fully filled.
func (r OrderResult) IsFilled() bool { return r.Status == StatusFilled }

// IsPartial reports whether the order is partially filled.
func (r OrderResult) IsPartial() bool { return r.Status == StatusPartiallyFilled }

// OpenOrder represents a resting order on the book. Mirrors order.py:OpenOrder.
type OpenOrder struct {
	Coin       string
	OrderID    int64
	Side       string // "B" buy, "S" sell
	Price      decimal.Decimal
	Size       decimal.Decimal
	Timestamp  int64
	OrderType  string
	ReduceOnly bool
}

// IsBuy reports whether this is a buy order.
func (o OpenOrder) IsBuy() bool { return o.Side == "B" }

// Position represents an open position. Size is positive for long and negative
// for short. Mirrors order.py:Position.
type Position struct {
	Coin          string
	Size          decimal.Decimal
	EntryPrice    decimal.Decimal
	PositionValue decimal.Decimal
	UnrealizedPnl decimal.Decimal
	Leverage      int
	LeverageType  string // "cross" or "isolated"
	LiquidationPx *decimal.Decimal

	// Risk tracking fields.
	MarkPrice      *decimal.Decimal
	FundingAccrued decimal.Decimal
	StopLossPrice  *decimal.Decimal
	OpenedAt       *time.Time

	// Calculated risk fields, populated by the risk manager.
	RiskAmount       *decimal.Decimal
	RiskPct          *decimal.Decimal
	CorrelationGroup *string
}

// IsLong reports whether the position is long (size > 0).
func (p Position) IsLong() bool { return p.Size.Sign() > 0 }

// IsShort reports whether the position is short (size < 0).
func (p Position) IsShort() bool { return p.Size.Sign() < 0 }

// AbsSize returns the absolute position size.
func (p Position) AbsSize() decimal.Decimal { return p.Size.Abs() }
